### ISSUE TITLE
Does gVisor support transparent huge pages?

### ISSUE THREAD

**Issue by RayWHL** [google/gvisor#8734]
### Description

Hello,

I'm wondering if gVisor supports the use of transparent huge pages. Specifically, I'm curious if a transparent huge page will be allocated if an application inside a gVisor container performs an mmap and triggers a page fault, given that the host machine has transparent huge pages enabled.

I've conducted a simple test comparing the performance of mmap and memory access inside a gVisor container with and without transparent huge pages enabled on the host machine. However, I found that there was almost no difference in performance.

Could you please help me confirm whether gVisor supports transparent huge pages or not?

**Comment by amscanne** [comment]
Yes, gVisor supports transparent huge pages.

This has a few parts:

* First, pgalloc ensures proper hugepage alignment for sufficiently large allocations [1].
* Second, VMAs are aligned on a hugepage basis where possible [2].
* Finally, PMAs are allocated on a hugepage basis as well [3].

Whether those pages are actually mapped as hugepages depends on the system's THP settings and hugepage availability. I'd be curious to know details of the test.

[1] https://cs.opensource.google/gvisor/gvisor/+/master:pkg/sentry/pgalloc/pgalloc.go;l=440?q=hugepage%20file:pgalloc&ss=gvisor
[2] https://cs.opensource.google/gvisor/gvisor/+/refs/heads/master:pkg/sentry/mm/vma.go;l=192;drc=2ec9f07a9d3bbb37100c8fa31b1b55b5a602b213;bpv=1;bpt=1
[3] https://cs.opensource.google/gvisor/gvisor/+/master:pkg/sentry/mm/pma.go;l=530?q=hugepage&ss=gvisor%2Fgvisor:pkg%2Fsentry%2F

**Comment by RayWHL** [comment]
Thank you for your response.

Here are the testing details:

First, I allocated 16GB of anonymous memory using mmap and accessed each 4KB page once to trigger a page fault, testing its execution time. Here's an example of the code:

```
buffer = mmap(NULL, NUM_PAGES * PAGE_SIZE, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
for (i = 0; i < NUM_PAGES * PAGE_SIZE; i += PAGE_SIZE) {
    buffer[i] = 'a';
    sum += buffer[i];
}
```

The testing process was as follows:

1. Enabled the host machine's transparent huge page mechanism:
```
echo always > /sys/kernel/mm/transparent_hugepage/enabled
```
2. Perform the test in gvisor.
3. Disable the host machine's transparent huge page mechanism.
```
echo never > /sys/kernel/mm/transparent_hugepage/enabled
```
4. Perform the test in gvisor again.

Result:  The tests showed that the execution time was roughly the same for both cases. However, when I enabled and disabled the transparent huge page mechanism directly on the host machine and performed the same test, there was a significant performance difference (approximately 3 times faster on thp).

It seems that gvisor did not benefit from thp. Did I miss any configuration?

My gVisor version:
```
runsc version release-20230313.0
spec: 1.1.0-rc.1
```

**Comment by RayWHL** [comment]
Thank you for your response.

Here are the testing details:

First, I allocated 16GB of anonymous memory using mmap and accessed each 4KB page once to trigger a page fault, testing its execution time. Here's an example of the code:

```
buffer = mmap(NULL, NUM_PAGES * PAGE_SIZE, PROT_READ | PROT_WRITE, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
for (i = 0; i < NUM_PAGES * PAGE_SIZE; i += PAGE_SIZE) {
    buffer[i] = 'a';
    sum += buffer[i];
}
```

The testing process was as follows:

1. Enabled the host machine's transparent huge page mechanism:
```
echo always > /sys/kernel/mm/transparent_hugepage/enabled
```
2. Perform the test in gvisor.
3. Disable the host machine's transparent huge page mechanism.
```
echo never > /sys/kernel/mm/transparent_hugepage/enabled
```
4.Perform the test in gvisor again.

Result:  The tests showed that the execution time was roughly the same for both cases. However, when I enabled and disabled the transparent huge page mechanism directly on the host machine and performed the same test, there was a significant performance difference (approximately 3 times faster on thp).

It seems that gvisor did not benefit from thp. Did I miss any configuration?

My gVisor version:
```
runsc version release-20230313.0
spec: 1.1.0-rc.1
```
